[NNC Testing] Randomized loop nest infrastructure

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70174

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33234529](https://our.internmc.facebook.com/intern/diff/D33234529)

anijain2305 committed:
[NNC Testing] Randomized loop nest infrastructure

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

anijain2305 commented:
Remaining comments from @navahgar from https://github.com/pytorch/pytorch/pull/64774

* Add a test to ensure that new changes don't break the infrastructure.
* Better readability for `no_more_inlining`

anijain2305 committed:
Update on "[NNC Testing] Randomized loop nest infrastructure"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

anijain2305 committed:
Update on "[NNC Testing] Randomized loop nest infrastructure"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

anijain2305 commented:
@ZolotukhinM @navahgar Can you PTAL?

anijain2305 commented:
@anijain2305 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33234529).

anijain2305 committed:
Update on "[NNC Testing] Randomized loop nest infrastructure"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33234529](https://our.internmc.facebook.com/intern/diff/D33234529)

[ghstack-poisoned]

anijain2305 commented:
@anijain2305 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33234529).

ZolotukhinM commented on test/test_jit_fuser_randomized.py:
```diff
@@ -0,0 +1,41 @@
+
+        # Set the seed to 1. This tests the codepath through random
+        # transformation.
+        os.environ["PYTORCH_TENSOREXPR_RANDOM_TRANSFORM_SEED"] = "1"
```
Does it not affect other tests running in CI? I.e. can this cause random transforms to be enabled in other tests that happen to run after this?

anijain2305 replied:
Oh I did not think of that possibility. Let me try and see if there is any better way.

anijain2305 replied:
Added `setUp` and `tearDown` functions, so it should alter the environment only during the specific test.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -3390,6 +3391,732 @@ bool LoopNest::rfactor(
 
+namespace randomization_helper {
+
+int64_t max_transformations(int n_max_transforms) {
+  // This env variable helps control the max number of transformations. This can
+  // be helpful in gradually reducing the number of transformations when we see
+  // an error.
+  static const char* enable_c_str =
+      std::getenv("PYTORCH_TENSOREXPR_RANDOM_MAX_TRANSFORM");
+  if (!enable_c_str) {
+    return n_max_transforms;
+  }
+  return std::stoi(std::string(enable_c_str));
+}
```
Can we use existing `PYTORCH_JIT_OPT_LIMIT` mechanism instead? It kinda does the same thing.

https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md#jit-optimization-limiter

anijain2305 replied:
Thanks. Moved the random loop nest transformations to a new file so that I can set up the env variable correctly. This also helps in better separation of code.

anijain2305 replied:
Added `setUp` and `tearDown` functions, so it should alter the environment only during the specific test.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -3390,6 +3391,732 @@ bool LoopNest::rfactor(
+  std::default_random_engine random_engine(seed);
+  std::srand(seed);
+  int n_transforms =
+      randomization_helper::max_transformations(std::rand() % 20);
```
Nit: let me put a clang-tidy hat: "magic constant '20'!"

anijain2305 replied:
handled

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -3390,6 +3391,732 @@ bool LoopNest::rfactor(
+    SLICE_HEAD, // 20
+    SLICE_TAIL,
+    CACHE_ACCESSES,
+    COMPUTE_AT, // better message
+    RFACTOR, // TODO
```
Do you mind expanding or removing these comments? Hard to tell what they mean

anijain2305 replied:
Done

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -3390,6 +3391,732 @@ bool LoopNest::rfactor(
+  };
+  bool can_inline = true;
+  try {
+    // n_transforms = 1;
```
Leftover?

anijain2305 replied:
Done

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -3390,6 +3391,732 @@ bool LoopNest::rfactor(
+    // n_transforms = 1;
+    for (int i = 0; i < n_transforms; i++) {
+      int transform = std::rand() % MAX_TRANSFORM;
+      // transform = seed;
```
And this

anijain2305 replied:
Done

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.h:
```diff
@@ -558,6 +558,8 @@ class TORCH_API LoopNest {
   }
   std::vector<BufPtr> getIntermediateBufs() const;
 
+  void randomTransform(int64_t seed);
```
Please add a comment for this function.

anijain2305 replied:
Done

ZolotukhinM reviewed (approved):
Thanks, overall looks good! I left some comments below.

anijain2305 committed:
Update on "[NNC Testing] Randomized loop nest infrastructure"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33234529](https://our.internmc.facebook.com/intern/diff/D33234529)

[ghstack-poisoned]

anijain2305 commented:
@anijain2305 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33234529).

anijain2305 committed:
Update on "[NNC Testing] Randomized loop nest infrastructure"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33234529](https://our.internmc.facebook.com/intern/diff/D33234529)

[ghstack-poisoned]

anijain2305 commented:
@anijain2305 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33234529).

anijain2305 committed:
Update on "[NNC Testing] Randomized loop nest infrastructure"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33234529](https://our.internmc.facebook.com/intern/diff/D33234529)

[ghstack-poisoned]

anijain2305 commented:
@anijain2305 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33234529).

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.h:
```diff
@@ -558,6 +558,15 @@ class TORCH_API LoopNest {
   std::vector<BufPtr> getIntermediateBufs() const;
 
+  // Finds which is the outer for loop between a and b for loops. Returns
+  // nullptr if they don't have same parent.
```
This function returns `nullptr` when neither of the given 2 Fors is an ancestor of the other. Not having the same parent is not being checked here. Actually, if the given 2 Fors have a common parent, this will return `nullptr`, since in that case, neither can be a parent of the other.

anijain2305 replied:
Thanks for catching the bad documentation. Updated it.

anijain2305 committed:
Update on "[NNC Testing] Randomized loop nest infrastructure"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33234529](https://our.internmc.facebook.com/intern/diff/D33234529)

[ghstack-poisoned]

anijain2305 commented on torch/csrc/jit/passes/loopnest_randomization.cpp:
```diff
@@ -0,0 +1,759 @@
+
+} // namespace randomization_helper
+
+void LoopNest::loopnestRandomization(int64_t seed) {
```
You are right. Refactored to add it as a separate file. PTAL.

anijain2305 commented:
@anijain2305 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33234529).

navahgar reviewed (approved):
LGTM. Thanks for refactoring it.

anijain2305 committed:
Update on "[NNC Testing] Randomized loop nest infrastructure"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33234529](https://our.internmc.facebook.com/intern/diff/D33234529)

[ghstack-poisoned]

anijain2305 commented:
@anijain2305 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33234529).
